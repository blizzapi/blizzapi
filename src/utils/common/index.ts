import constants from '../../constants';
import {
  validateRegionId,
  validateRegionName,
  getRegionIdByName,
} from '../localization/regions';
import {
  ConstantKey,
  RegionIdOrName,
  RegionIdAsNumberOrString,
  RegionName,
} from '../../types';

export const getConstantByRegionId = (
  regionId: RegionIdAsNumberOrString,
  constantKey: ConstantKey,
) => {
  const regionIdAsString = typeof regionId !== 'string'
    ? regionId
    : regionId.toString();
  const isRegionIdValid = validateRegionId(regionIdAsString);

  if (!isRegionIdValid) {
    throw new RangeError(
      `${regionIdAsString} is not a valid parameter for getConstantByRegionId(${regionId}, '${constantKey}')`,
    );
  }
  return constants[constantKey][regionIdAsString];
};

/**
 * Returns constant value for given region name from specified constant object
 *
 * @param regionName Region name as two-letter string (e.g. 'us')
 * @param constantKey Constant key
 * @return Uri as a string
 */
export const getConstantByRegionName = (regionName: RegionName, constantKey: ConstantKey) => {
  const isRegionNameValid = validateRegionName(regionName);

  if (!isRegionNameValid) {
    throw new RangeError(
      `${regionName} is not a valid parameter for getConstantByRegionName(${regionName}, '${constantKey}')`,
    );
  }

  const regionId = getRegionIdByName(regionName);

  return constants[constantKey][regionId];
};

export const getConstantByRegion = (regionIdOrName: RegionIdOrName, constantKey: ConstantKey) =>
  (validateRegionId(regionIdOrName)
    ? getConstantByRegionId(regionIdOrName, constantKey)
    : getConstantByRegionName(regionIdOrName.toString(), constantKey));
