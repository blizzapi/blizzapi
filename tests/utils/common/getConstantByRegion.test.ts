import { ConstantKey } from '../../../@types';
import * as utils from '../../../src/lib/utils/common/index';

const { getConstantByRegion } = utils;

import regionNamesJson from '../../__testData__/regionNames.json';
import wrongRegionNamesJson from '../../__testData__/wrongRegionNames.json';
import regionIdsJson from '../__testData__/index';
import wrongRegionIdsJson from '../../__testData__/wrongRegionIds.json';
import constantKeysJson from '../../__testData__/constantKeys.json';

/* tslint:disable no-expression-statement */
describe('getConstantByRegion()', () => {
  test('should be defined', () => {
    expect(getConstantByRegion).toBeDefined();
  });

  test('should be of type "function"', () => {
    expect(typeof getConstantByRegion).toBe('function');
  });

  (regionIdsJson as ReadonlyArray<number | string>).forEach(regionId =>
    (constantKeysJson as ReadonlyArray<ConstantKey>).forEach(constantKey => {
      test(`should return correct value for ${regionId} and ${constantKey} as valid parameters`, () => {
        expect(getConstantByRegion(regionId, constantKey)).toMatchSnapshot();
      });
    }),
  );

  (wrongRegionIdsJson as ReadonlyArray<number | string>).forEach(wrongRegionId =>
    (constantKeysJson as ReadonlyArray<ConstantKey>).forEach(constantKey => {
      test(`should throw RangeError for ${wrongRegionId} as invalid region id`, () => {
        expect(() => getConstantByRegion(wrongRegionId, constantKey)).toThrow(RangeError);
      });
    }),
  );

  (regionNamesJson as ReadonlyArray<string>).forEach(regionName =>
    (constantKeysJson as ReadonlyArray<ConstantKey>).forEach(constantKey => {
      test(`should return correct value for ${regionName} as valid region name`, () => {
        expect(getConstantByRegion(regionName, constantKey)).toMatchSnapshot();
      });
    }),
  );

  (wrongRegionNamesJson as ReadonlyArray<string>).forEach(wrongRegionName =>
    (constantKeysJson as ReadonlyArray<ConstantKey>).forEach(constantKey => {
      test(`should throw RangeError for ${wrongRegionName} as invalid region name`, () => {
        expect(() => getConstantByRegion(wrongRegionName, constantKey)).toThrow(RangeError);
      });
    }),
  );
});
/* tslint:enable no-expression-statement */
