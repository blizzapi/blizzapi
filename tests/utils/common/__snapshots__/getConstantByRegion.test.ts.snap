// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getConstantByRegion() should return correct value for 1 and OAUTH_AUTHORIZE_URIS as valid parameters 1`] = `"https://us.battle.net/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for 1 and OAUTH_AUTHORIZE_URIS as valid parameters 2`] = `"https://us.battle.net/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for 1 and OAUTH_CHECK_TOKEN_URIS as valid parameters 1`] = `"https://us.battle.net/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for 1 and OAUTH_CHECK_TOKEN_URIS as valid parameters 2`] = `"https://us.battle.net/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for 1 and OAUTH_TOKEN_URIS as valid parameters 1`] = `"https://us.battle.net/oauth/token"`;

exports[`getConstantByRegion() should return correct value for 1 and OAUTH_TOKEN_URIS as valid parameters 2`] = `"https://us.battle.net/oauth/token"`;

exports[`getConstantByRegion() should return correct value for 1 and REGION_API_HOSTS as valid parameters 1`] = `"https://us.api.blizzard.com"`;

exports[`getConstantByRegion() should return correct value for 1 and REGION_API_HOSTS as valid parameters 2`] = `"https://us.api.blizzard.com"`;

exports[`getConstantByRegion() should return correct value for 2 and OAUTH_AUTHORIZE_URIS as valid parameters 1`] = `"https://eu.battle.net/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for 2 and OAUTH_AUTHORIZE_URIS as valid parameters 2`] = `"https://eu.battle.net/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for 2 and OAUTH_CHECK_TOKEN_URIS as valid parameters 1`] = `"https://eu.battle.net/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for 2 and OAUTH_CHECK_TOKEN_URIS as valid parameters 2`] = `"https://eu.battle.net/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for 2 and OAUTH_TOKEN_URIS as valid parameters 1`] = `"https://eu.battle.net/oauth/token"`;

exports[`getConstantByRegion() should return correct value for 2 and OAUTH_TOKEN_URIS as valid parameters 2`] = `"https://eu.battle.net/oauth/token"`;

exports[`getConstantByRegion() should return correct value for 2 and REGION_API_HOSTS as valid parameters 1`] = `"https://eu.api.blizzard.com"`;

exports[`getConstantByRegion() should return correct value for 2 and REGION_API_HOSTS as valid parameters 2`] = `"https://eu.api.blizzard.com"`;

exports[`getConstantByRegion() should return correct value for 3 and OAUTH_AUTHORIZE_URIS as valid parameters 1`] = `"https://apac.battle.net/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for 3 and OAUTH_AUTHORIZE_URIS as valid parameters 2`] = `"https://apac.battle.net/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for 3 and OAUTH_CHECK_TOKEN_URIS as valid parameters 1`] = `"https://apac.battle.net/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for 3 and OAUTH_CHECK_TOKEN_URIS as valid parameters 2`] = `"https://apac.battle.net/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for 3 and OAUTH_TOKEN_URIS as valid parameters 1`] = `"https://apac.battle.net/oauth/token"`;

exports[`getConstantByRegion() should return correct value for 3 and OAUTH_TOKEN_URIS as valid parameters 2`] = `"https://apac.battle.net/oauth/token"`;

exports[`getConstantByRegion() should return correct value for 3 and REGION_API_HOSTS as valid parameters 1`] = `"https://kr.api.blizzard.com"`;

exports[`getConstantByRegion() should return correct value for 3 and REGION_API_HOSTS as valid parameters 2`] = `"https://kr.api.blizzard.com"`;

exports[`getConstantByRegion() should return correct value for 5 and OAUTH_AUTHORIZE_URIS as valid parameters 1`] = `"https://www.battlenet.com.cn/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for 5 and OAUTH_AUTHORIZE_URIS as valid parameters 2`] = `"https://www.battlenet.com.cn/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for 5 and OAUTH_CHECK_TOKEN_URIS as valid parameters 1`] = `"https://www.battlenet.com.cn/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for 5 and OAUTH_CHECK_TOKEN_URIS as valid parameters 2`] = `"https://www.battlenet.com.cn/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for 5 and OAUTH_TOKEN_URIS as valid parameters 1`] = `"https://www.battlenet.com.cn/oauth/token"`;

exports[`getConstantByRegion() should return correct value for 5 and OAUTH_TOKEN_URIS as valid parameters 2`] = `"https://www.battlenet.com.cn/oauth/token"`;

exports[`getConstantByRegion() should return correct value for 5 and REGION_API_HOSTS as valid parameters 1`] = `"https://api.blizzard.com.cn"`;

exports[`getConstantByRegion() should return correct value for 5 and REGION_API_HOSTS as valid parameters 2`] = `"https://api.blizzard.com.cn"`;

exports[`getConstantByRegion() should return correct value for CN as valid region name 1`] = `"https://api.blizzard.com.cn"`;

exports[`getConstantByRegion() should return correct value for CN as valid region name 2`] = `"https://www.battlenet.com.cn/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for CN as valid region name 3`] = `"https://www.battlenet.com.cn/oauth/token"`;

exports[`getConstantByRegion() should return correct value for CN as valid region name 4`] = `"https://www.battlenet.com.cn/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for Cn as valid region name 1`] = `"https://api.blizzard.com.cn"`;

exports[`getConstantByRegion() should return correct value for Cn as valid region name 2`] = `"https://www.battlenet.com.cn/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for Cn as valid region name 3`] = `"https://www.battlenet.com.cn/oauth/token"`;

exports[`getConstantByRegion() should return correct value for Cn as valid region name 4`] = `"https://www.battlenet.com.cn/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for EU as valid region name 1`] = `"https://eu.api.blizzard.com"`;

exports[`getConstantByRegion() should return correct value for EU as valid region name 2`] = `"https://eu.battle.net/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for EU as valid region name 3`] = `"https://eu.battle.net/oauth/token"`;

exports[`getConstantByRegion() should return correct value for EU as valid region name 4`] = `"https://eu.battle.net/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for Eu as valid region name 1`] = `"https://eu.api.blizzard.com"`;

exports[`getConstantByRegion() should return correct value for Eu as valid region name 2`] = `"https://eu.battle.net/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for Eu as valid region name 3`] = `"https://eu.battle.net/oauth/token"`;

exports[`getConstantByRegion() should return correct value for Eu as valid region name 4`] = `"https://eu.battle.net/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for KR as valid region name 1`] = `"https://kr.api.blizzard.com"`;

exports[`getConstantByRegion() should return correct value for KR as valid region name 2`] = `"https://apac.battle.net/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for KR as valid region name 3`] = `"https://apac.battle.net/oauth/token"`;

exports[`getConstantByRegion() should return correct value for KR as valid region name 4`] = `"https://apac.battle.net/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for Kr as valid region name 1`] = `"https://kr.api.blizzard.com"`;

exports[`getConstantByRegion() should return correct value for Kr as valid region name 2`] = `"https://apac.battle.net/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for Kr as valid region name 3`] = `"https://apac.battle.net/oauth/token"`;

exports[`getConstantByRegion() should return correct value for Kr as valid region name 4`] = `"https://apac.battle.net/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for TW as valid region name 1`] = `"https://kr.api.blizzard.com"`;

exports[`getConstantByRegion() should return correct value for TW as valid region name 2`] = `"https://apac.battle.net/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for TW as valid region name 3`] = `"https://apac.battle.net/oauth/token"`;

exports[`getConstantByRegion() should return correct value for TW as valid region name 4`] = `"https://apac.battle.net/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for Tw as valid region name 1`] = `"https://kr.api.blizzard.com"`;

exports[`getConstantByRegion() should return correct value for Tw as valid region name 2`] = `"https://apac.battle.net/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for Tw as valid region name 3`] = `"https://apac.battle.net/oauth/token"`;

exports[`getConstantByRegion() should return correct value for Tw as valid region name 4`] = `"https://apac.battle.net/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for Us as valid region name 1`] = `"https://us.api.blizzard.com"`;

exports[`getConstantByRegion() should return correct value for Us as valid region name 2`] = `"https://us.battle.net/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for Us as valid region name 3`] = `"https://us.battle.net/oauth/token"`;

exports[`getConstantByRegion() should return correct value for Us as valid region name 4`] = `"https://us.battle.net/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for cN as valid region name 1`] = `"https://api.blizzard.com.cn"`;

exports[`getConstantByRegion() should return correct value for cN as valid region name 2`] = `"https://www.battlenet.com.cn/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for cN as valid region name 3`] = `"https://www.battlenet.com.cn/oauth/token"`;

exports[`getConstantByRegion() should return correct value for cN as valid region name 4`] = `"https://www.battlenet.com.cn/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for cn as valid region name 1`] = `"https://api.blizzard.com.cn"`;

exports[`getConstantByRegion() should return correct value for cn as valid region name 2`] = `"https://www.battlenet.com.cn/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for cn as valid region name 3`] = `"https://www.battlenet.com.cn/oauth/token"`;

exports[`getConstantByRegion() should return correct value for cn as valid region name 4`] = `"https://www.battlenet.com.cn/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for eU as valid region name 1`] = `"https://eu.api.blizzard.com"`;

exports[`getConstantByRegion() should return correct value for eU as valid region name 2`] = `"https://eu.battle.net/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for eU as valid region name 3`] = `"https://eu.battle.net/oauth/token"`;

exports[`getConstantByRegion() should return correct value for eU as valid region name 4`] = `"https://eu.battle.net/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for eu as valid region name 1`] = `"https://eu.api.blizzard.com"`;

exports[`getConstantByRegion() should return correct value for eu as valid region name 2`] = `"https://eu.battle.net/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for eu as valid region name 3`] = `"https://eu.battle.net/oauth/token"`;

exports[`getConstantByRegion() should return correct value for eu as valid region name 4`] = `"https://eu.battle.net/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for kR as valid region name 1`] = `"https://kr.api.blizzard.com"`;

exports[`getConstantByRegion() should return correct value for kR as valid region name 2`] = `"https://apac.battle.net/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for kR as valid region name 3`] = `"https://apac.battle.net/oauth/token"`;

exports[`getConstantByRegion() should return correct value for kR as valid region name 4`] = `"https://apac.battle.net/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for kr as valid region name 1`] = `"https://kr.api.blizzard.com"`;

exports[`getConstantByRegion() should return correct value for kr as valid region name 2`] = `"https://apac.battle.net/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for kr as valid region name 3`] = `"https://apac.battle.net/oauth/token"`;

exports[`getConstantByRegion() should return correct value for kr as valid region name 4`] = `"https://apac.battle.net/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for tW as valid region name 1`] = `"https://kr.api.blizzard.com"`;

exports[`getConstantByRegion() should return correct value for tW as valid region name 2`] = `"https://apac.battle.net/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for tW as valid region name 3`] = `"https://apac.battle.net/oauth/token"`;

exports[`getConstantByRegion() should return correct value for tW as valid region name 4`] = `"https://apac.battle.net/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for tw as valid region name 1`] = `"https://kr.api.blizzard.com"`;

exports[`getConstantByRegion() should return correct value for tw as valid region name 2`] = `"https://apac.battle.net/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for tw as valid region name 3`] = `"https://apac.battle.net/oauth/token"`;

exports[`getConstantByRegion() should return correct value for tw as valid region name 4`] = `"https://apac.battle.net/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for uS as valid region name 1`] = `"https://us.api.blizzard.com"`;

exports[`getConstantByRegion() should return correct value for uS as valid region name 2`] = `"https://us.battle.net/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for uS as valid region name 3`] = `"https://us.battle.net/oauth/token"`;

exports[`getConstantByRegion() should return correct value for uS as valid region name 4`] = `"https://us.battle.net/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for us as valid region name 1`] = `"https://us.api.blizzard.com"`;

exports[`getConstantByRegion() should return correct value for us as valid region name 2`] = `"https://us.battle.net/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for us as valid region name 3`] = `"https://us.battle.net/oauth/token"`;

exports[`getConstantByRegion() should return correct value for us as valid region name 4`] = `"https://us.battle.net/oauth/check_token?token="`;

exports[`getConstantByRegion() should return correct value for us as valid region name 5`] = `"https://us.api.blizzard.com"`;

exports[`getConstantByRegion() should return correct value for us as valid region name 6`] = `"https://us.battle.net/oauth/authorize"`;

exports[`getConstantByRegion() should return correct value for us as valid region name 7`] = `"https://us.battle.net/oauth/token"`;

exports[`getConstantByRegion() should return correct value for us as valid region name 8`] = `"https://us.battle.net/oauth/check_token?token="`;
